### DISCUSSION TITLE
Relationship picker not retrieving all records

### DISCUSSION THREAD

**Discussion by JCV67** [Budibase/budibase#11558] [Help] [answered]
In a form, I use Relationship picker to choose clients from the clients table. I have no filters on that. Now, on picker, I can only see the first clients (Until those started by B). I suspect that is has something to to with the new filter ability.

On the datasouce I can see all records .

Choosing the reccord from data source table, results on showing all the clients on the design tables.

Just found out that the Relationship Picker is fetching only 100 records.

self hosted | v2.9.25 | Internal tables

**Comment by melohagan** [comment]
Hey @JCV67 

There is a hard limit of 100 options that will be returned by the relationship picker. Is it possible to apply a filter to get a subset of relevant options?

**Comment by melohagan** [comment]
Hey @JCV67 

There is a hard limit of 100 options that will be returned by the relationship picker. Is it possible to apply a filter to get a subset of relevant options?

**Reply by JCV67** [reply]
Hey @melohagan 
That wouldn't be a problem if the 'autocomplete' search box would search from within the entire table. But this way, relationship tables are useless for me. All was working well until a few days ago. 
I need to pick a client, any client, from clients related table, is there any way of doing that?
Thank you
JCV

**Reply by melohagan** [reply]
Hey @JCV67 You can use a data provider as the options source of an options picker: https://docs.budibase.com/docs/option-picker#using-a-data-provider

**Reply by JCV67** [reply]
What will I do with all the existing records,  and the automation that sends a report to the client. I would need to query other datasource.

**Reply by melohagan** [reply]
How many options would you need, about 1000 or so? If you're dealing with this many rows, maybe there's another approach you could take, such as using some block component that has a search field.

**Reply by JCV67** [reply]
This is a Maintenance Tasks App. Today I have 1045 rows on clients datasouce. My main concern is that changing the way I relate maintenance tasks records with clients would cause data loss on existing records. Why did you enforced a cap on data picker?

**Reply by melohagan** [reply]
A cap was enforced on the picker because you can't really expect a user to scroll through a thousand options, although your point is noted about the search functionality.

**Reply by JCV67** [reply]
Yes, just need the search ability.

**Comment by JCV67** [comment]
Hi @melohagan, do you intend to do something about this issue? The way relationship picker is working now will force me to make big changes an my apps, this is a big issue to me. Appreciate your help.

**Reply by ConorWebb96** [reply, marked as answer]
Hey @JCV67,

You likely will need to refactor in this case. I've created an enhancement here based around this here; https://github.com/Budibase/budibase/issues/11607. However, this is unlikely to get priority so it may be sometime before this gets added. (This is why I'm saying it would be best to refactor in this case.)

Hopefully this makes sense.

**Comment by JCV67** [comment]
Hello @ConorWebb96  thank you for your reply
This is a big issue to me. The rules changed in the middle of the game. I need to find out another way to pick a record from another table while filling a form.

**Comment by jmlaurenti** [comment]
HI @melohagan and @ConorWebb96 ,
I'm afraid I'm suffering from the same issue..
I have build an asset management system (basically employee list and computers/monitors list)... we are about 130 employees, and have about 160 PC... so since latest change, I just cannot access full list of computers when using the computers picker from the employee form.
I can understand displaying only 100, but only if the search capability of the picker is able to return the requested item..
Like other guys above, it's an unexpected change of behavior that may lead to significant refactor :-( 

As always, thanks for the hard work !

Best regards,
JML

**Reply by ConorWebb96** [reply]
Hey @jmlaurenti,

Would this plugin be helpful for you? https://github.com/ConorWebb96/bb-type-ahead It would allow you to search through a specific table and then bind your data. If i remember correctly you can load up to 1000 records by changing the limit whilst using an external data source. E.g. MySQL Postgres etc.

**Reply by JCV67** [reply]
Hi @jmlaurentim, thank you, 

Some of my tables have more than 1000 rows and in two of my main apps I use internal tables. Also I need to pick records while filling a form. So it would not solve my problem. I really don't no what to do.

**Reply by jmlaurenti** [reply]
@ConorWebb96 Just played with your plugin, it's working great for my use case ! So far I can wait for the BB fix to avoid significant changes in my code, but I might consider using it in some scenarios...
Thank ! :-)

**Comment by pierreboudes** [comment]
Hi @JCV67 and @jmlaurenti,
I have the same issue and I think it would be a bad idea to revert to an older version of BB. I hope that the #11639 of @melohagan  will find its way through production quickly. In the meantime I am loosing users, who will surely go back to a shared spreadsheet. Have you find any way to patch your app while keeping th current version of BB ? I though of increasing that hard coded value but my deployment uses docker compose so this isn't really trivial (or is it ?).   

I have tried your plugin @ConorWebb96, but I didn't figure how to use it in my context: I need to add contacts in a table (postgres) each one attached to an organisation (another table) so I "just" need to pick the right foreign key by name. 

Sorry, I am still a bit lost with BB, I though it would be easier to use it than to code my modest app myself.

**Reply by jmlaurenti** [reply]
Hi @pierreboudes , had no choice but to revert to older BBversion for the moment. Just read comment below from @JCV67, seems a fix/improvement is on its way ! :-)
Thanks @ConorWebb96 and @melohagan for quick action :-)

Best regards, 
JML

**Comment by JCV67** [comment]
I think this problem will be solved   [BUDI-7426] Relationship picker searching enhancement #11607 
Hope it does not take to long.

**Reply by pierreboudes** [reply]
The merge/pull request is in the pipes #11639 but such things usually take time.

**Comment by JCV67** [comment]
Dommage

**Comment by JCV67** [comment]
Hello @melohagan @ConorWebb96 and every one who discussed this subject. Relationship picker is now working fine with the new "pre-search" filter. It retrieves the needed records and it is fast. Thank you very much. Just one little thing I couldn't find a way of customizing the place holder witch is "Search by 'column label'" 'Search by' is in English and the column label isn't always readable by users. A small tweak on this would be great. Anyway it is working very well.  Great!!!

**Reply by melohagan** [reply]
Great to hear! If you wouldn't mind @JCV67 could you raise a feature request for the search by placeholder label if that's something you'd be interested in

**Comment by likke812** [comment]
Hello @melohagan,

I upgrade my Budibase to the latest version(2.10.1) and your search enhancement for relationship picker is working fine until it combined with a filter which caused the search is not working?

I need to insert a row to table epg while on the Budibase application it needs to do relationship picker - "vlan_id" from another vlan table which has 4094 records. And during the relationship pickup, we'd like to filter out those "vlan.status == Used" and only let the user choose "vlan.status == Free" records.

But if I add this filter set at the Relationship Picker -> Settings -> Filtering, then the pre-searching for relationship picker will not work. Before adding the filter, the search function for relationship picker is working fine.

Is there anything rule I missed or it has a bug on it?

Thank you!

**Reply by melohagan** [reply]
Hey @likke812 If a filter is provided then the search will be client side (with the record limit of 100), it only does a server side search when no filter is provided.

If you are using relationship picker filters, then you can uncheck the *Search* setting of the picker, and then add a form with text field to do the searching.

Similar to this: https://docs.budibase.com/docs/adding-a-search-field#2-using-a-form-and-text-input-flexible-way

**Reply by likke812** [reply]
Thanks for the quick response. From your post, I understand this search enhancement can not compatible with Filtering so I will have  to add a new form with a 123 field and a table to filter out what I need for vlan_id and status pair. Then manually choice them at relationship picker.

**Comment by JCV67** [comment]
Hello @melohagan 

Sorry for the delay, here's a video of what's happening to me:

https://github.com/Budibase/budibase/assets/104722906/2745f902-bc5b-4d39-a723-2d986f143929

**Reply by melohagan** [reply]
Hey @JCV67 Thanks, I had re-opened this issue a while back: https://github.com/Budibase/budibase/issues/11478#issuecomment-1805899243

**Reply by JCV67** [reply]
thank you, just noted that my comment was made on the wrong discussion.
